spec/services/session_encryptor_spec.rb

```diff
@@ -21,6 +21,18 @@
     expect(encryptor2.load(encrypted_text)).to eq(payload)
   end
 
+  it 'does not modify the cek when failing to decrypt a payload encrypted with an old key' do
+    old_encryptor = SessionEncryptor.new
+    old_encryptor.user_access_key.unlock(Pii::Cipher.random_key)
+    old_payload = old_encryptor.dump('asdf' => '1234')
+
+    new_encryptor = SessionEncryptor.new
+    original_cek = new_encryptor.user_access_key.cek
+
+    expect { new_encryptor.load(old_payload) }.to raise_error(Pii::EncryptionError)
```

Comment:
Does this mean we'll start seeing these encryption errors? I don't understand what this is testing.

Reply:
I was wondering the same. Does a user with a session created from the old code get the `Pii::EncryptionError` on their next request? Do we have any easy way to handle that if so?

Reply:
We won't start seeing them. We're seeing them now, and they're rescued [here](https://github.com/18F/redis-session-store/blob/master/lib/redis-session-store.rb#L106) in the `redis-session-store` gem. That deletes the session which is why we're seeing CSRF errors.

Reply:
Is the reason we're seeing them now due to the old way the SessionEncryptor class was written? If so, is there a way to test that with this new code, we won't be seeing these errors anymore?

Reply:
We've been seeing them all along. That's the reason for the CSRF error. If an old session is decrypted before the SessionEncryptor's user access key is unlocked, then the user access key will assume the `random_r` of the old session. Since the user access key is not unlocked at init, that is effectively a random value.

What we want to see, is decryption errors when trying to decrypt old sessions that were unlocked with the invalid and indeterminate `random_r`. Those sessions will be deleted and then recreated with the new determinate `random_r`.

Reply:
I understand now. Thanks for the explanation.

Reply:
I had a thought. We could dup and lock the access key right before decryption? That adds some overhead to every request, but, it:

- Avoids re-calculating the scrypt hash
- Allows us to decrypt old sessions, avoiding raising this error
- Allows non-deterministic user access keys

Maybe something to investigate in a follow-on PR though since the current state is everything is on 🔥

Reply:
> * Avoids re-calculating the scrypt hash

That would be great. If I recall that is one of our biggest performance hits so it sounds like that would half our response time for some requests.

Reply:
To be clear, we aren't re-calculating the scrypt hash now, since we are memoizing the user access key.

Reply:
Would it make sense to add logging to the `redis-session-store` gem so that we can see when errors happen?

Reply:
That is probably a good idea. May be useful upstream as well.